### PR TITLE
CGE-106 Possibility to logout

### PR DESCRIPTION
**Pull request by drw85**
On "/auth/logout" current users token will be placed to ConcurrentHashMap in InvalidatedTokensHolder class. Logic of token validation was extended - now all authorized requests are checking for absence in the array of invalidated tokens.

### CODE REVIEW

File: src/main/java/com/hillel/items_exchange/config/AppConfig.java [hunk 1]
```diff
@@ -6,13 +6,15 @@
...
 import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
+import org.springframework.scheduling.annotation.EnableScheduling;
 import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
 import org.springframework.validation.beanvalidation.LocalValidatorFactoryBean;
 import org.springframework.web.servlet.LocaleResolver;
 import org.springframework.web.servlet.i18n.AcceptHeaderLocaleResolver;
 
 @Configuration
 @EnableJpaAuditing(dateTimeProviderRef = "dateTimeProvider")
+@EnableScheduling
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/config/AppConfig.java:17]: Really good solution!

File: src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.util.Date;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class InvalidatedTokensHolder {
+
+    private final ConcurrentHashMap<String, Date> data = new ConcurrentHashMap<>();
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java:16]: I remind you about using a map interface here.

File: src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    public void invalidate(final String token, final Date expireDate) {
+        data.put(token, expireDate);
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java:19]: Use computeIfAbsent here it's makes opperation more concurrent secure.

File: src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenFilter.java [hunk 4]
```diff
@@ -16,13 +16,16 @@
...
     private final JwtTokenProvider jwtTokenProvider;
+    private final InvalidatedTokensHolder invalidatedTokensHolder;
 
     @Override
     public void doFilter(ServletRequest req, ServletResponse res, FilterChain filterChain)
             throws IOException, ServletException {
 
         String token = jwtTokenProvider.resolveToken((HttpServletRequest) req);
-        if (!token.isEmpty() && jwtTokenProvider.validateToken(token, (HttpServletRequest) req)) {
+        if (!token.isEmpty()
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenFilter.java:26]: Could you extract the condition to a new method with more informal name?

File: src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenProvider.java [hunk 5]
```diff
@@ -87,4 +89,19 @@ public boolean validateToken(String token, HttpServletRequest req) {
...
+
+    public void invalidateToken(String token) {
+        final Date expirationDate = getTokenExpirationDate(token)
+                .orElseThrow(() -> new JwtException(getExceptionMessageSource("invalid.token")));
+        invalidatedTokensHolder.invalidate(token, expirationDate);
+    }
+
+    public Optional<Date> getTokenExpirationDate(String token) {
+        try {
+            return Optional.of(Jwts.parser().setSigningKey(secret).parseClaimsJws(token).getBody().getExpiration());
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenProvider.java:101]: I just want to clearify that Optiona.of doesn't expect null value. Did you aware of it when created the code?

File: src/test/java/com/hillel/items_exchange/controller/AuthControllerIntegrationTest.java [hunk 6]
```diff
@@ -35,6 +38,8 @@
 
     @Autowired
     private MockMvc mockMvc;
+    @Autowired
+    InvalidatedTokensHolder invalidatedTokensHolder;
```
Comment (rpkyrych) [inline src/test/java/com/hillel/items_exchange/controller/AuthControllerIntegrationTest.java:42]: Would you mind to use the team's checkstyle and to add private modifier here?

File: src/test/java/com/hillel/items_exchange/controller/AuthControllerIntegrationTest.java [hunk 7]
```diff
@@ -152,4 +157,37 @@ void login_Success_shouldReturnHttpOk() throws Exception {
...
+        final String token = obtainToken(createUserLoginDto());
+        mockMvc.perform(post(LOGOUT_URL)
+                .header("Authorization", "Bearer " + token)
+                .contentType(MediaType.APPLICATION_JSON)
+                .accept(MediaType.APPLICATION_JSON))
+                .andDo(print());
+        Assert.assertTrue(invalidatedTokensHolder.isInvalidated(token));
+    }
+
+    private String obtainToken(UserLoginDto loginDto) throws Exception {
```
Comment (rpkyrych) [inline src/test/java/com/hillel/items_exchange/controller/AuthControllerIntegrationTest.java:185]: Nice approach!

**Review by rpkyrych — CHANGES_REQUESTED** [review]

**Review by Jack11M — APPROVED** [review]
Great job!

**Review by lubchik-14 — CHANGES_REQUESTED** [review]
Excellent job, but there is few test scenario.

File: src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        data.computeIfAbsent(token, v -> expireDate);
+    }
+
+    public boolean isInvalidated(final String token) {
+        return data.containsKey(token);
+    }
+
+
+    @Scheduled(fixedDelayString = "${app.jwt.expiration.time.ms}")
+    private void removeExpiredTokens() {
```
[the remark this answers is not in the document]
Comment (drw85) [reply src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java:29]: Why? This methods periodically deletes only the expired one.

File: src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenFilter.java [hunk 9]
```diff
@@ -16,19 +16,24 @@
...
             Authentication auth = jwtTokenProvider.getAuthentication(token);
-
             if (auth != null) {
                 SecurityContextHolder.getContext().setAuthentication(auth);
             }
         }
         filterChain.doFilter(req, res);
     }
+
+    private boolean isTokenValidAndNotBlacklisted(HttpServletRequest req, String token) {
```
[the remark this answers is not in the document]
Comment (drw85) [reply src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenFilter.java:34]: This logic puts tokens in some sort of blacklist after logout. Seems good name for me)))

File: src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenFilter.java [hunk 10]
```diff
@@ -16,19 +16,24 @@
...
+        return !token.isEmpty()
+                && jwtTokenProvider.validateToken(token, req)
+                && !invalidatedTokensHolder.isInvalidated(token);
```
[the remark this answers is not in the document]
Comment (drw85) [reply src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenFilter.java:37]: Agree. Good approach. Thanks.

File: src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenProvider.java [hunk 11]
```diff
@@ -87,4 +89,19 @@ public boolean validateToken(String token, HttpServletRequest req) {
             return false;
         }
     }
+
+    public void invalidateToken(String token) {
+        final Date expirationDate = getTokenExpirationDate(token)
+                .orElseThrow(() -> new JwtException(getExceptionMessageSource("invalid.token")));
+        invalidatedTokensHolder.invalidate(token, expirationDate);
```
[the remark this answers is not in the document]
Comment (drw85) [reply src/main/java/com/hillel/items_exchange/security/jwt/JwtTokenProvider.java:96]: Thanks. My bad)

File: src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Date;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class InvalidatedTokensHolder {
+
+    private final ConcurrentMap<String, Date> data = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (drw85) [reply src/main/java/com/hillel/items_exchange/security/jwt/InvalidatedTokensHolder.java]: You are right. Thnx.

File: src/main/java/com/hillel/items_exchange/config/SecurityConfig.java [hunk 13]
```diff
@@ -27,6 +28,7 @@
     public static final String HAS_ROLE_USER = "hasRole('ROLE_USER')";
     private final JwtTokenProvider jwtTokenProvider;
     private final JwtAuthenticationEntryPoint jwtAuthenticationEntryPoint;
+    private final InvalidatedTokensHolder invalidatedTokensHolder;
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/config/SecurityConfig.java:31]: Unused bean.

File: src/main/java/com/hillel/items_exchange/controller/AuthController.java [hunk 14]
```diff
@@ -82,6 +83,13 @@
         }
     }
 
+    @PostMapping("/logout")
+    public ResponseEntity<HttpStatus> logout(HttpServletRequest req){
+        final String token = jwtTokenProvider.resolveToken(req);
+        jwtTokenProvider.invalidateToken(token);
+        return new ResponseEntity<>(HttpStatus.OK);
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/controller/AuthController.java:90]: You could make the method void and use annotation ResponseStatus(HttpStatus.OK) but it up to you.

**Review by rpkyrych — APPROVED** [review]

**Review by lubchik-14 — APPROVED** [review]

File: src/main/java/com/hillel/items_exchange/config/SecurityConfig.java (hunk 13 above)
Comment (drw85) [reply src/main/java/com/hillel/items_exchange/config/SecurityConfig.java:31]: I've tried to extract it but @RequiredArgsConstructor said that he needs it.

**Pull request merged**